### ISSUE TITLE
Microsoft Dynamics 365 contact entity and birthdate field (DateTime issue)

### ISSUE THREAD

**Issue by jamesmacwhite** [verbb/formie#1265]
### Describe the bug

So I ended up going down a bit of an Dynamics 365 API hole, but did get to the bottom of an potential caveat to the DateTime handling in the main integration class.

Pushing a DateTime value to the `birthdate` field on the contact entity which is a standard field can potentially yield a 400 error from the API.

This is the entity data for reference:

```json
 {
    "@odata.type": "#Microsoft.Dynamics.CRM.DateTimeAttributeMetadata",
    "AttributeType": "DateTime",
    "IsCustomAttribute": false,
    "IsValidForCreate": true,
    "IsValidForUpdate": true,
    "CanBeSecuredForCreate": true,
    "CanBeSecuredForUpdate": true,
    "LogicalName": "birthdate",
    "SchemaName": "BirthDate",
    "MetadataId": "9fb79f93-3449-4dec-a6b6-96faf3fdc0c7",
    "DisplayName": {
        "LocalizedLabels": [
            {
                "Label": "Date of Birth",
                "LanguageCode": 1033,
                "IsManaged": true,
                "MetadataId": "6140b506-2341-db11-898a-0007e9e17ebd",
                "HasChanged": null
            }
        ],
        "UserLocalizedLabel": {
            "Label": "Date of Birth",
            "LanguageCode": 1033,
            "IsManaged": true,
            "MetadataId": "6140b506-2341-db11-898a-0007e9e17ebd",
            "HasChanged": null
        }
    },
    "RequiredLevel": {
        "Value": "None",
        "CanBeChanged": true,
        "ManagedPropertyLogicalName": "canmodifyrequirementlevelsettings"
    }
}
```

A very minimal payload test on the contact entity, the datetime value is a mock up of what is currently provided by the default integration when using a date field without time specifically enabled:

```json
{
  "firstname": "James",
  "lastname": "White",
  "birthdate": "2000-12-12 00:00:00"
}
```

I got a 400 bad request with the first payload test, which is what the integration class provides, as the AttributeType is DateTime per the EntityDefinition.

> Error identified in Payload provided by the user for Entity :'contacts', For more information on this error please follow this help link https://go.microsoft.com/fwlink/?linkid=2195293  ---->  InnerException : Microsoft.OData.ODataException: Cannot convert the literal '2000-12-12 00:00:00' to the expected type 'Edm.Date'. ---> System.FormatException: String '2000-12-12 00:00:00' was not recognized as a valid Edm.Date.

This however, works fine.

```json
{
  "firstname": "James",
  "lastname": "White",
  "birthdate": "2000-12-12"
}
```

So interestingly the error mentions Edm.Date not Edm.DateTime which is apparently what this entity should be based on the EntityDefinition, but by basically dropping `00:00:00` it works fine.

Now per the base integration as the AttributeType is stated as DateTime, this will cause an issue because a time value will be added, given it matches the DateTime type condition. In order to workaround the problem, I can override the type for the specific handle and make it appear as just `date` so this means it gets formatted with date value instead and works, but this seemed odd and I dug deeper.

Testing other DateTime attribute type fields does not seem to have issues with DateTime, so there seemed to be something specific with birthdate. I then found some Microsoft documentation that suggests `DateOnly` fields exist and birthdate is one of them given it is a standard out of the box field.

https://learn.microsoft.com/en-us/dynamics365/customerengagement/on-premises/customize/behavior-format-date-time-field?view=op-9-1

While the docs state it's for CRM on premises environments, I think the core information is actually general to Microsoft Dynamics 365 so ignore that. On the EntityDefinition there is another property called "DateTimeBehavior", this however is currently not included in the Entity Definitions query as just metadata properties are selected, but if you view all without any `$select` you can see the property. On birthdate it has the value `DateOnly`, which hints that while it's set as DateTime, it's not actually accepting DateTime, just to be super difficult.

So there's potentially further date handling required for fields that are set as the AttributeType DateTime but actually in reality only accept Date values only. The document linked provides the default fields on various entities that are indeed DateOnly, however this does show a scenario where you cannot trust the AttributeType entirely specifically with DateTime.

Using a specific EntityDefinition query for DateTime attributes using something like this:

```
GET EntityDefinitions(LogicalName='$entity')/Attributes/Microsoft.Dynamics.CRM.DateTimeAttributeMetadata?$select=SchemaName,LogicalName,DateTimeBehavior
```

That then gives you all the DateTime fields for that entity and what the DateTimeBehavior is.

Looking at the birthdate field:

```json
{
    "LogicalName": "birthdate",
    "MetadataId": "9fb79f93-3449-4dec-a6b6-96faf3fdc0c7",
    "DateTimeBehavior": {
        "Value": "DateOnly"
    }
},
```

And now we have revealed that the troublesome birthdate field that while claims it is DateTime by it's AttributeType, doesn't actually mean it will allow you to use DateTime values, given DateOnly is set!

In the context of the entities this plugin supports, all of these are DateOnly fields by default. Try pushing a DateTime value to any of these fields and the API is going to tell you to do one.

**Contact:**

- Anniversary
- Birthdate

**Lead:** 

- Est. Close Date

**Opportunity:**

- Actual Close Date
- Est. Close Date
- Final Decision Date

So the integration class would need to set the type to just as `IntegrationField::TYPE_DATE` for such fields, but this not obvious without further EntityDefinition interrogation specifically on DateTime AttributeType fields.

While it is an additional EntityDefinition query, doing this would also be able to handle this for custom date fields in addition to the standard ones that may also be DateOnly and you could possibly even further filter the EntityDefinition query itself to just to return DateOnly DateTime fields, given that's all we need to get to make sure you don't send a DateTime when not supported.

Then it is just a case of checking through the field handles that are DateOnly values and then setting the `$type` to `Date` for the `_convertFieldType()` function to assign the type correctly. The base integration class would now need to include Date in addition to DateTime though which it currently does not, as by AttributeType "Date" is never seen because Microsoft don't use it.

Shout out to Microsoft for making this one more complicated than it needs to be. Maybe just use a different AttributeType?!

### Steps to reproduce

1. Test DateTime value with the brithdate field on contact entity.
2. Get a Edm.Date error due to not supporting the time part of the date.
3. Go down a Microsoft Dynamics 365 hole for a while on why a DateTime field is complaining about a valid DateTime value
4. Come back up from said hole and realise Microsoft just wants to watch the world burn, but at least feel better you have an explanation for point 3!

### Form settings

N/A

### Craft CMS version

3.7.63.1

### Plugin version

1.6.21 (dev-craft-3)

### Multi-site?

No

**Comment by engram-design** [comment]
Thanks for investigating, and yep using `IntegrationField::TYPE_DATE` to describe date-only fields is what we want.

**Comment by jamesmacwhite** [comment]
Yes, it does mean we have to be more defensive on DateTime fields and check for DateOnly before assigning the type, however querying the specific DateTime attribute data should resolve it in #1263. Unfortunately, we can't get DateTimeBehavior in the existing Metadata request, because it's specific to `Microsoft.Dynamics.CRM.DateTimeAttributeMetadata` so it's another GET, but a fairly small one and does at least prevent the possibility of a DateTime value being sent to a DateOnly field.

**Comment by jamesmacwhite** [comment]
Resolved by #1263
